### PR TITLE
Implement core template loader overrides to rely on wp_template posts

### PR DESCRIPTION
**Pull request by felixarntz**
## Description
Based on #17513, this PR overrides the WordPress core template loader mechanism to rely on `wp_template` posts instead of theme template files. It fully maintains the existing WordPress template hierarchy ecosystem players are already familiar with.

A similar approach was previously explored in #4659, however with a slightly different implementation.

## Types of changes
* Record existing template hierarchy to find a matching `wp_template` post (i.e. block template). This is a drastic change from how themes work today and therefore highly experimental.
* Use a template canvas / skeleton to print basic markup (`html`, `head` tags, `wp_head` action invokation, etc.).

### Related considerations
* Wrap block template content in `div.wp-site-blocks` to allow for specific descendant styling such as `.wp-site-blocks > *`. Without this, you could of course do `body > *`, but this may unexpectedly affect other markup printed via `wp_footer` for example.
* Disallow deletion of the `wp_template` post with slug `index`, as this should always exist (eventually this could be auto-generated if it doesn't exist yet, based on the current theme).
* While this PR is basically usable in itself, it only becomes valuable with the most essential full-site block types being available (see #15623 for context). A `wp_template` post is also registered by other PRs (e.g. #17263) because they rely on the most basic functionality available. I believe that we should merge both these most essential block types and this `wp_template` foundation before we make the functionality available to a user in any way (i.e. before we expose the "Full Site Editing" feature in the "Experimental" section).

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: lib/template-loader.php
Comment (westonruter) [inline lib/template-loader.php]:
```suggestion
		$_wp_current_template_post = array_shift( $template_posts );
```

File: lib/template-loader.php
Comment (westonruter) [inline lib/template-loader.php]: Is this necessary? Couldn't the `body` be used instead of `.wp-site-blocks`? Alternatively, the `wp-site-blocks` class could be added to the `body_classes` when a block-based template is being served as opposed to a PHP-based template.

File: lib/template-loader.php
Comment (westonruter) [inline lib/template-loader.php]: I know this is in `the_content()` but why is it?

File: lib/template-loader.php
Comment (westonruter) [inline lib/template-loader.php]: Why only `wptexturize()`? Should there rather be a new filter applied like `wp_template_content` which then can have filters as core applies:

```php
add_filter( 'wp_template_content', 'do_blocks', 9 );
add_filter( 'wp_template_content', 'wptexturize' );
add_filter( 'wp_template_content', 'convert_smilies', 20 );
add_filter( 'wp_template_content', 'wpautop' );
add_filter( 'wp_template_content', 'shortcode_unautop' );
add_filter( 'wp_template_content', 'prepend_attachment' );
add_filter( 'wp_template_content', 'wp_make_content_images_responsive' );
add_filter( 'wp_template_content', 'do_shortcode', 11 ); // AFTER wpautop()
add_filter( 'wp_template_content', 'capital_P_dangit', 11 );
```

File: lib/template-canvas.php
Comment (epiqueras) [reply lib/template-canvas.php]: Wow, great catch!

File: lib/template-loader.php
Comment (epiqueras) [reply lib/template-loader.php]: `body` will also style `wp_footer` content.

File: lib/template-loader.php
Comment (epiqueras) [reply lib/template-loader.php]: This might be very useful for theme devs.

cc @youknowriad

File: lib/template-loader.php
Comment (epiqueras) [inline lib/template-loader.php]: When will this ever run more than once and require this merging?

And what guarantees subsequent runs pass `$templates` where the first item is of lesser priority than the last item of the current `$_wp_current_template_hierarchy`?

File: lib/template-loader.php
Comment (felixarntz) [reply lib/template-loader.php]: Exactly, in a way I agree it'd be nicer to not have this extra div, but the way many themes handle design rules for block content is to use a selector like `.{blocks-wrap-class} > *`, which would indeed style `wp_footer` content that it shouldn't apply to.

I'd say let's be conservative for now and start with this div. We should definitely continue to look for an alternative solution though that satisfies the styling needs.

File: lib/template-loader.php
Comment (felixarntz) [reply lib/template-loader.php]: That is worth a discussion, actually about two things:
1. Do we want to introduce a filter here?
2. Which functions should the content be run through by default?

Regarding 1., I would prefer to start without one and only introduce it based on requests that are reasonable. The `the_content` filter is mostly misused to append/prepend additional content, which in a block-based world becomes unnecessary - having a `wp_template_content` being misused in the same way would most likely introduce broader issues (since it covers the entire page markup).

Regarding 2., I think some of the filter functions definitely don't make sense for `wp_template` content (e.g. `prepend_attachment`). Others we need to discuss further (like do we want to deal with shortcodes at all in this?).

File: lib/template-loader.php
Comment (felixarntz) [reply lib/template-loader.php]: Based on how core's template loader works, this _may_ be run multiple times (e.g. `get_index_template()` being called after another `get_*_template()` function). As well based on core, the functions are called in the correct hierarchy order (more specific come first), so it will always be correct regarding priority.

Note that this is only a layer to make it work in the plugin anyway. Most likely, if/when this gets merged into core, it would be more directly integrated with the existing template loader code.

File: lib/template-loader.php
Comment (epiqueras) [reply lib/template-loader.php]: I agree with your points for `1`.

I am not familiar enough with these filters to comment about `2`. Maybe we should ping more people who worked on them?

File: lib/template-loader.php
Comment (epiqueras) [reply lib/template-loader.php]: Thanks, that makes sense.

**Comment by felixarntz** [comment]
@westonruter @epiqueras What do you think about this given the previous discussion? I wonder whether we can merge this as is for now, as I think the eventual discussion on whether we add a filter add what we run through it can be deferred, especially since this is going to be behind an experimental flag.

Is there anything else here you think we should cover?

**Comment by epiqueras** [comment]
@felixarntz Let's merge this as is after giving the new globals a more Gutenberg and file specific prefix to avoid any conflicts in the future.

The e2e tests also need to be fixed?

File: lib/template-loader.php
Comment (Copons) [reply lib/template-loader.php]: In the WP.com implementation of Full Site Editing templates we noticed that we had to run [several filters on the template content](https://github.com/Automattic/wp-calypso/blob/b4bafbc1818998d67661ffc62c41cf2fa09b8582/apps/full-site-editing/full-site-editing-plugin/full-site-editing/templates/class-wp-template.php#L180-L220).

We basically started adding them manually whenever we caught an issue, but then basically added all of them (see for example the useless `prepend_attachment` that I added—thanks for pointing it out @felixarntz! 😅) to save us headaches.
For example, I vividly recall how `wp_make_content_images_responsive` almost single-handedly fixed all image related issue we were having.

Comment (felixarntz) [reply lib/template-loader.php]: Thanks @Copons, I just found that code as well and wanted to comment here, but you beat me to it. :)

Based on that and your comment, I think what we should add to this PR for now are the following:
* `$wp_embed->run_shortcode( $content )`
* `$wp_embed->autoembed( $content )`
* `wp_make_content_images_responsive( $content )`

File: lib/template-loader.php
Comment (epiqueras) [reply lib/template-loader.php]: I think it's because `wptexturize` doesn't handle that specific case?

File: lib/template-loader.php
Comment (epiqueras) [reply lib/template-loader.php]: Done!

**Review by epiqueras — APPROVED** [review]
LGTM

**Comment by JoshuaJarman** [comment]
We are having issues with using these wp_template custom post types to override woocommerce templates.  any tips?

**Pull request merged**
